alphabase/psm_reader/psm_reader.py

```diff
@@ -219,30 +221,36 @@ def add_modification_mapping(self, modification_mapping: dict) -> None:
             ```
 
         """
-        if (
-            modification_mapping is None
-            or len(modification_mapping) == 0
-            or not isinstance(modification_mapping, dict)
-        ):
+        if not isinstance(modification_mapping, dict):
             return
 
+        new_modification_mapping = defaultdict(list)
         for key, val in list(modification_mapping.items()):
-            if key in self.modification_mapping:
-                if isinstance(val, str):
-                    self.modification_mapping[key].append(val)
-                else:
-                    self.modification_mapping[key].extend(val)
-            elif isinstance(val, str):
-                self.modification_mapping[key] = [val]
+            if isinstance(val, str):
+                new_modification_mapping[key].append(val)
             else:
-                self.modification_mapping[key] = val
+                new_modification_mapping[key].extend(val)
 
-        self.set_modification_mapping(self.modification_mapping)
+        if new_modification_mapping:
+            self.set_modification_mapping(
+                self.modification_mapping | new_modification_mapping
+            )
 
     def set_modification_mapping(
         self, modification_mapping: Optional[dict] = None
     ) -> None:
-        """Set the modification mapping."""
+        """Set the modification mapping for the search engine.
+
+        Also creates a reverse mapping from the modification format used by the search engine to the AlphaBase format.
+
+        Parameters
+        ----------
+        modification_mapping:
+            If dictionary: the current modification_mapping will be overwritten by this.
+            If str: the parameter will be interpreted as a reader type, and the modification_mapping is read from the
+                "modification_mapping" section of the psm_reader_yaml
+
+        """
         if modification_mapping is None:
             self._init_modification_mapping()
         elif isinstance(modification_mapping, str):
```

Comment:
@jalew188 here, a lot of hidden complexity arises (also connected to `_init_modification_mapping`) as the "column_mapping" parameter in the yaml is overloaded (could either be str or dict).
Could we introduce a (mutually exclusive ) "column_mapping_type` in the yaml, that served to reference another reader?

Reply:
Yes, we shall then add a new method: `add_new_column_mapping_item` to handle user-defined columns

Reply:
Both modification_mapping and column_mapping is not well designed in alphabase readers. We need a better way.

Reply:
I absolutely agree. I think the modification mapping is already quite close but we need to make sure it works the same where everywhere. It's also not yet compatible with custom mods and MP.